gzip_verilog.v.gz failing on macOS

## Steps to reproduce the issue

As part of https://github.com/NixOS/nixpkgs/pull/68107, I'm trying to get the tests to pass on macOS when built with Nix. Unfortunately, the gzip_verilog test [seems to be failing](https://dev.azure.com/arcnmx/nixexprs/_build/results?buildId=970) and I'm not sure why. Perhaps it's failing to link to `zlib`? But it's included in the build dependencies so there should be no problems there. Unfortunately I don't have direct access to a macOS machine so I can't tell whether this is a general macOS problem or something happening due to the Nix build environment. Any suggestions on how to narrow down this failure would be appreciated.

## Expected behavior

All tests pass, like they do on Linux.

## Actual behavior

```
Running gzip_verilog.ys..
gzip_verilog.v.gz:1: ERROR: syntax error, unexpected $undefined
make[1]: *** [run-test.mk:25: run-gzip_verilog.ys] Error 1
```

--- a/kernel/register.cc
+++ b/kernel/register.cc
@@ -48,7 +48,7 @@ using zlib to write gzip-compressed data every time the stream is flushed.
 */
 class gzip_ostream : public std::ostream  {
 public:
-	gzip_ostream()
+	gzip_ostream() : std::ostream(nullptr)
 	{
 		rdbuf(&outbuf);
 	}
@@ -71,7 +71,7 @@ private:
 			str("");
 			return 0;
 		}
-		~gzip_streambuf()
+		virtual ~gzip_streambuf()
 		{
 			sync();
 			gzclose(gzf);
@@ -498,7 +498,15 @@ void Frontend::extra_args(std::istream *&f, std::string &filename, std::vector<s
 			if (f != NULL) {
 				// Check for gzip magic
 				unsigned char magic[3];
-				int n = readsome(*ff, reinterpret_cast<char*>(magic), 3);
+				int n = 0;
+				while (n < 3)
+				{
+					int c = ff->get();
+					if (c != EOF) {
+						magic[n] = (unsigned char) c;
+					}
+					n++;
+				}
 				if (n == 3 && magic[0] == 0x1f && magic[1] == 0x8b) {
 	#ifdef YOSYS_ENABLE_ZLIB
 					log("Found gzip magic in file `%s', decompressing using zlib.\n", filename.c_str());

